# Notebook for parser development

<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Loading-data" data-toc-modified-id="Loading-data-1">Loading data</a></span></li><li><span><a href="#Check,-remove-duplicates" data-toc-modified-id="Check,-remove-duplicates-2">Check, remove duplicates</a></span></li><li><span><a href="#Column-level-transforms" data-toc-modified-id="Column-level-transforms-3">Column-level transforms</a></span></li><li><span><a href="#Pre-NodeNorming" data-toc-modified-id="Pre-NodeNorming-4">Pre-NodeNorming</a></span><ul class="toc-item"><li><span><a href="#Exploring:-Genes" data-toc-modified-id="Exploring:-Genes-4.1">Exploring: Genes</a></span><ul class="toc-item"><li><span><a href="#HGNC" data-toc-modified-id="HGNC-4.1.1">HGNC</a></span></li><li><span><a href="#OMIM" data-toc-modified-id="OMIM-4.1.2">OMIM</a></span></li><li><span><a href="#Comparing-HGNC-vs-OMIM" data-toc-modified-id="Comparing-HGNC-vs-OMIM-4.1.3">Comparing HGNC vs OMIM</a></span></li><li><span><a href="#Conclusions" data-toc-modified-id="Conclusions-4.1.4">Conclusions</a></span></li></ul></li><li><span><a href="#Exploring:-Diseases" data-toc-modified-id="Exploring:-Diseases-4.2">Exploring: Diseases</a></span><ul class="toc-item"><li><span><a href="#OMIM/orphanet" data-toc-modified-id="OMIM/orphanet-4.2.1">OMIM/orphanet</a></span></li><li><span><a href="#MONDO" data-toc-modified-id="MONDO-4.2.2">MONDO</a></span></li><li><span><a href="#Comparing-OMIM/orphanet-vs-MONDO" data-toc-modified-id="Comparing-OMIM/orphanet-vs-MONDO-4.2.3">Comparing OMIM/orphanet vs MONDO</a></span></li><li><span><a href="#Checking-MONDO-data" data-toc-modified-id="Checking-MONDO-data-4.2.4">Checking MONDO data</a></span></li><li><span><a href="#Conclusions" data-toc-modified-id="Conclusions-4.2.5">Conclusions</a></span></li></ul></li></ul></li><li><span><a href="#Stats-on-rows-removed-during-NodeNorming" data-toc-modified-id="Stats-on-rows-removed-during-NodeNorming-5">Stats on rows removed during NodeNorming</a></span></li><li><span><a href="#Adding-NodeNorm-data,-removing-rows" data-toc-modified-id="Adding-NodeNorm-data,-removing-rows-6">Adding NodeNorm data, removing rows</a></span></li><li><span><a href="#Confidence-values" data-toc-modified-id="Confidence-values-7">Confidence values</a></span><ul class="toc-item"><li><span><a href="#Removing-rows-+-stats" data-toc-modified-id="Removing-rows-+-stats-7.1">Removing rows + stats</a></span></li></ul></li><li><span><a href="#Generating-documents" data-toc-modified-id="Generating-documents-8">Generating documents</a></span></li><li><span><a href="#Checking-documents" data-toc-modified-id="Checking-documents-9">Checking documents</a></span></li><li><span><a href="#Parser-notes" data-toc-modified-id="Parser-notes-10">Parser notes</a></span></li></ul></div>

In [1]:
## not for parser. for notebook only 

## CX: allows multiple lines of code to print from one code block
from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = "all"

## Loading data

__Current approach__ is to load all files into 1 pandas dataframe. Then I can...

1. check the duplicates situation: records found in multiple panel files. I can check whether the same record looks different between files or not (by checking duplicates using all columns vs key columns). -> Raise errors if yes
2. remove duplicates before generating documents
3. Do some tasks column-wise over all the data, rather than while iterating over rows

Notes:
* There are a few existing BioThings parsers that also use `pandas` to load the entire raw data file at once: https://github.com/search?q=repo%3Abiothings%2Fpending.api%20pandas&type=code
* But there are other parsers that use `csv` to load the file **one row at a time** (generator): https://github.com/search?q=repo%3Abiothings%2Fpending.api+csv+reader&type=code

---

If I did the generator approach (load files 1 by 1, 1 row at a time), I'd have to modify how I do things:
1. Don't do the duplicates check. But try to mitigate potential "duplicate" issues: 
   - Sort all delimited strings
   - Use a hash of all column values (when they're all strings) for `_id`. Want rows with all the same values to produce the same hash
2. Either leave to BioThings toolset to remove duplicates, or could save a running set of `_id` hashes to check if row was already encountered -> not create duplicate docs
3. Do the tasks on single rows/chunks (pandas [read_csv](https://pandas.pydata.org/docs/reference/api/pandas.read_csv.html#pandas.read_csv) has an iterator for rows/chunks! see iterator/chunksize parameters)

In [2]:
## put into parser: DONE
import pathlib
import pandas as pd

## don't put in parser. Just for this notebook
import glob
from pprint import pprint

## unsure on putting into parser: more for notebook viewing/debugging...
pd.options.display.max_columns = None

<div class="alert alert-block alert-danger">

Adjust the code block below for path/pattern for data files. 
    
This notebook was originally written using data files from the 2025-02-28 release to FTP site 

In [3]:
## put into parser (format): DONE

base_file_path = pathlib.Path.home().joinpath("Desktop", "EBIgene2pheno_files", 
                                              "From_FTP", "2025-03-28")

## uses pathlib's Path.glob, which produces a generator. 
## cast into list so parser code can check if paths were actually matched or not
all_file_paths = list(base_file_path.glob("*.csv.gz"))
all_file_paths

[PosixPath('/Users/colleenxu/Desktop/EBIgene2pheno_files/From_FTP/2025-03-28/DDG2P_2025-03-28.csv.gz'),
 PosixPath('/Users/colleenxu/Desktop/EBIgene2pheno_files/From_FTP/2025-03-28/HearingLossG2P_2025-03-28.csv.gz'),
 PosixPath('/Users/colleenxu/Desktop/EBIgene2pheno_files/From_FTP/2025-03-28/SkinG2P_2025-03-28.csv.gz'),
 PosixPath('/Users/colleenxu/Desktop/EBIgene2pheno_files/From_FTP/2025-03-28/CancerG2P_2025-03-28.csv.gz'),
 PosixPath('/Users/colleenxu/Desktop/EBIgene2pheno_files/From_FTP/2025-03-28/CardiacG2P_2025-03-28.csv.gz'),
 PosixPath('/Users/colleenxu/Desktop/EBIgene2pheno_files/From_FTP/2025-03-28/SkeletalG2P_2025-03-28.csv.gz'),
 PosixPath('/Users/colleenxu/Desktop/EBIgene2pheno_files/From_FTP/2025-03-28/EyeG2P_2025-03-28.csv.gz')]

In [4]:
## an example: pathlib's Path.glob produces a generator
## vs glob.glob produces an array (from cwd only?)
base_file_path.glob("*2025-02-28.csv.gz")
glob.glob("*2025-02-28.csv.gz")

<generator object Path.glob at 0x1131e3010>

[]

In [5]:
## put into parser (format): DONE

## using generator expression (think list/dict comprehension) within pd.concat to load files 1 at a time
## ingesting all columns as str for now
df = pd.concat((pd.read_csv(f, dtype=str) for f in all_file_paths), ignore_index=True)

## make column names snake-case - usable with itertuples later
df.columns = df.columns.str.replace(" ", "_")

In [6]:
df["date_of_last_review"].info(memory_usage="deep")

<class 'pandas.core.series.Series'>
RangeIndex: 4726 entries, 0 to 4725
Series name: date_of_last_review
Non-Null Count  Dtype 
--------------  ----- 
4726 non-null   object
dtypes: object(1)
memory usage: 341.7 KB


In [7]:
## change this column to datetime, saves memory
df["date_of_last_review"] = pd.to_datetime(df["date_of_last_review"])
df["date_of_last_review"].info(memory_usage="deep")

<class 'pandas.core.series.Series'>
RangeIndex: 4726 entries, 0 to 4725
Series name: date_of_last_review
Non-Null Count  Dtype              
--------------  -----              
4726 non-null   datetime64[ns, UTC]
dtypes: datetime64[ns, UTC](1)
memory usage: 37.1 KB


In [8]:
## I couldn't figure out how to import + ingest column as datetime in 1 step 
## this is what I tried that didn't work

## worked with pandas 2.0.3, but didn't work with pandas 2.2.3: ingested as str
# df = pd.concat((pd.read_csv(f, dtype=str, parse_dates=["date of last review"]) 
#                 for f in all_file_paths), ignore_index=True)

## doesn't work
# df = pd.concat((pd.read_csv(f, dtype=str, parse_dates=["date of last review"], 
#                            date_format="%Y-%m-%d %H:%M:%S%:z") 
#                 for f in all_file_paths), ignore_index=True)
## throws an error
# df = pd.concat((pd.read_csv(f, dtype=str, parse_dates=[["date of last review"]], 
#                            date_format="%Y-%m-%d %H:%M:%S%:z") 
#                 for f in all_file_paths), ignore_index=True)
## throws an error
# df = pd.concat((pd.read_csv(f, dtype={"date of last review": pd.datetime64[ns, tz]})
#                 for f in all_file_paths), ignore_index=True)

In [9]:
df.shape
df.head()

(4726, 21)

g2p_id gene_symbol gene_mim hgnc_id              previous_gene_symbols  \
0  G2P00001        HMX1   142992    5017                         H6; NKX5-3   
1  G2P00002        SLX4   613278   23845  BTBD12; FANCP; KIAA1784; KIAA1987   
2  G2P00003        ARG1   608313     663                                NaN   
3  G2P00004         ATR   601215     882            FRP1; MEC1; SCKL; SCKL1   
4  G2P00005       FANCB   300515    3583              FAAP95; FAB; FLJ34064   

                           disease_name disease_mim disease_MONDO  \
0  HMX1-related oculoauricular syndrome      612109           NaN   
1           SLX4-related Fanconi anemia      613951           NaN   
2              ARG1-related argininemia      207800           NaN   
3           ATR-related Seckel syndrome      210600           NaN   
4          FANCB-related Fanconi anemia      300514           NaN   

        allelic_requirement cross_cutting_modifier  confidence  \
0       biallelic_autosomal                    NaN  definitive   
1       biallelic_autosomal                    NaN  definitive   
2       biallelic_autosomal                    NaN  definitive   
3       biallelic_autosomal                    NaN      strong   
4  monoallelic_X_hemizygous                    NaN  definitive   

   variant_consequence variant_types molecular_mechanism  \
0  absent gene product           NaN    loss of function   
1  absent gene product           NaN    loss of function   
2  absent gene product           NaN    loss of function   
3  absent gene product           NaN    loss of function   
4  absent gene product           NaN    loss of function   

  molecular_mechanism_categorisation molecular_mechanism_evidence  \
0                           inferred                          NaN   
1                           inferred                          NaN   
2                           inferred                          NaN   
3                           inferred                          NaN   
4                           inferred                          NaN   

                                          phenotypes  \
0  HP:0000568; HP:0000589; HP:0000007; HP:0000482...   
1  HP:0000347; HP:0000007; HP:0001903; HP:0002984...   
2  HP:0000752; HP:0000737; HP:0000007; HP:0008339...   
3  HP:0000347; HP:0010230; HP:0001249; HP:0002750...   
4  HP:0000924; HP:0001871; HP:0001701; HP:0000083...   

                                        publications         panel comments  \
0                       18423520; 25574057; 29140751       DD; Eye      NaN   
1                                 21240275; 21240277            DD      NaN   
2       10502833; 1598908; 7649538; 1463019; 2365823            DD      NaN   
3                                                NaN  DD; Skeletal      NaN   
4  16679491; 21910217; 36135330; 32106311; 307922...      DD; Skin      NaN   

        date_of_last_review  
0 2019-09-26 16:23:46+00:00  
1 2025-01-28 23:09:54+00:00  
2 2015-07-22 16:14:07+00:00  
3 2025-01-27 14:24:27+00:00  
4 2024-08-20 14:13:58+00:00

In [10]:
df.info(memory_usage="deep")

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4726 entries, 0 to 4725
Data columns (total 21 columns):
 #   Column                              Non-Null Count  Dtype              
---  ------                              --------------  -----              
 0   g2p_id                              4726 non-null   object             
 1   gene_symbol                         4726 non-null   object             
 2   gene_mim                            4724 non-null   object             
 3   hgnc_id                             4726 non-null   object             
 4   previous_gene_symbols               4251 non-null   object             
 5   disease_name                        4726 non-null   object             
 6   disease_mim                         3576 non-null   object             
 7   disease_MONDO                       641 non-null    object             
 8   allelic_requirement                 4726 non-null   object             
 9   cross_cutting_modifier              632 n

## Check, remove duplicates

There are duplicate rows in this dataframe because the record (gene + disease + more) is in several panels (disease falls into multiple categories). This was explored in the data-playground notebook. 

We want to drop those duplicates. 
However, I was concerned that the delimited-string values could differ (only in list order) for the same record in diff files. 
So that's what this check is for. 

In [11]:
## put into parser (format): DONE

n_duplicates_column_combo = df[df.duplicated(subset=["g2p_id"], keep=False)].shape

n_duplicates_all_columns = df[df.duplicated(keep=False)].shape

## for testing
# n_duplicates_all_columns = (1, 1)


if n_duplicates_column_combo != n_duplicates_all_columns: 
    raise AssertionError("The data format has changed, and the assumptions about duplicates/key columns may " \
                          "no longer hold. Re-explore the data and adjust the parser.")

In [12]:
## put into parser (format): DONE

## drop duplicates
df.drop_duplicates(inplace=True, ignore_index=True)

In [13]:
df.info(memory_usage="deep")

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3659 entries, 0 to 3658
Data columns (total 21 columns):
 #   Column                              Non-Null Count  Dtype              
---  ------                              --------------  -----              
 0   g2p_id                              3659 non-null   object             
 1   gene_symbol                         3659 non-null   object             
 2   gene_mim                            3657 non-null   object             
 3   hgnc_id                             3659 non-null   object             
 4   previous_gene_symbols               3287 non-null   object             
 5   disease_name                        3659 non-null   object             
 6   disease_mim                         2572 non-null   object             
 7   disease_MONDO                       564 non-null    object             
 8   allelic_requirement                 3659 non-null   object             
 9   cross_cutting_modifier              454 n

## Column-level transforms

Based on data-playground "Notes on parsing data to create documents" section

In [14]:
## double-checking how to add prefixes to OMIM vs orphanet IDs

df_diseasemim = df.copy()

## done to preserve NA
df_diseasemim["disease_mim"] = [i if pd.isna(i) \
                                else "OMIM:" + i if i.isnumeric() \
                                else i \
                                for i in df_diseasemim["disease_mim"]]

df_diseasemim["disease_mim"] = df_diseasemim["disease_mim"].str.replace("Orphanet", "orphanet")

In [15]:
df_diseasemim[df_diseasemim["disease_mim"].str.contains("OMIM:", na=False)].shape

df_diseasemim[df_diseasemim["disease_mim"].str.contains("orphanet:", na=False)].shape

## add up row count. If == num non-null in info above, you're good 
## right now 2570 == 2570, so good

(2563, 21)

(9, 21)

In [16]:
## put into parser (format): DONE

## COLUMN-LEVEL TRANSFORMS

## adding Translator/biolink prefixes to IDs
df["gene_mim"] = "OMIM:" + df["gene_mim"]
df["hgnc_id"] = "HGNC:" + df["hgnc_id"]
df["disease_mim"] = df["disease_mim"].str.replace("Orphanet", "orphanet")
## done to preserve NA
df["disease_mim"] = [i if pd.isna(i)
                     else "OMIM:" + i if i.isnumeric()
                     else i
                     for i in df["disease_mim"]]

## strip whitespace
df["disease_name"] = df["disease_name"].str.strip()
df["comments"] = df["comments"].str.strip()

## create new columns
## UI really wants resource website urls like this. May need to adjust over time as website changes
df["g2p_record_url"] = "https://www.ebi.ac.uk/gene2phenotype/lgd/" +  df["g2p_id"]

## replace panel keywords with full names shown on G2P website for single record
## keeping "Hearing loss" as-is, changing all other values
df["panel"] = df["panel"].str.replace("DD", "Developmental disorders")
df["panel"] = df["panel"].str.replace("Cancer", "Cancer disorders")
df["panel"] = df["panel"].str.replace("Cardiac", "Cardiac disorders")
df["panel"] = df["panel"].str.replace("Eye", "Eye disorders")
df["panel"] = df["panel"].str.replace("Skeletal", "Skeletal disorders")
df["panel"] = df["panel"].str.replace("Skin", "Skin disorders")

In [17]:
## checking on column-level transforms

df.head()
# df["g2p record url"].unique()[0:100]

# df[df["disease mim"].str.contains("orphanet", na=False)]  ## 9 rows, so that's correct
# df[df["panel"].str.contains("Hearing", na=False)]

g2p_id gene_symbol     gene_mim     hgnc_id  \
0  G2P00001        HMX1  OMIM:142992   HGNC:5017   
1  G2P00002        SLX4  OMIM:613278  HGNC:23845   
2  G2P00003        ARG1  OMIM:608313    HGNC:663   
3  G2P00004         ATR  OMIM:601215    HGNC:882   
4  G2P00005       FANCB  OMIM:300515   HGNC:3583   

               previous_gene_symbols                          disease_name  \
0                         H6; NKX5-3  HMX1-related oculoauricular syndrome   
1  BTBD12; FANCP; KIAA1784; KIAA1987           SLX4-related Fanconi anemia   
2                                NaN              ARG1-related argininemia   
3            FRP1; MEC1; SCKL; SCKL1           ATR-related Seckel syndrome   
4              FAAP95; FAB; FLJ34064          FANCB-related Fanconi anemia   

   disease_mim disease_MONDO       allelic_requirement cross_cutting_modifier  \
0  OMIM:612109           NaN       biallelic_autosomal                    NaN   
1  OMIM:613951           NaN       biallelic_autosomal                    NaN   
2  OMIM:207800           NaN       biallelic_autosomal                    NaN   
3  OMIM:210600           NaN       biallelic_autosomal                    NaN   
4  OMIM:300514           NaN  monoallelic_X_hemizygous                    NaN   

   confidence  variant_consequence variant_types molecular_mechanism  \
0  definitive  absent gene product           NaN    loss of function   
1  definitive  absent gene product           NaN    loss of function   
2  definitive  absent gene product           NaN    loss of function   
3      strong  absent gene product           NaN    loss of function   
4  definitive  absent gene product           NaN    loss of function   

  molecular_mechanism_categorisation molecular_mechanism_evidence  \
0                           inferred                          NaN   
1                           inferred                          NaN   
2                           inferred                          NaN   
3                           inferred                          NaN   
4                           inferred                          NaN   

                                          phenotypes  \
0  HP:0000568; HP:0000589; HP:0000007; HP:0000482...   
1  HP:0000347; HP:0000007; HP:0001903; HP:0002984...   
2  HP:0000752; HP:0000737; HP:0000007; HP:0008339...   
3  HP:0000347; HP:0010230; HP:0001249; HP:0002750...   
4  HP:0000924; HP:0001871; HP:0001701; HP:0000083...   

                                        publications  \
0                       18423520; 25574057; 29140751   
1                                 21240275; 21240277   
2       10502833; 1598908; 7649538; 1463019; 2365823   
3                                                NaN   
4  16679491; 21910217; 36135330; 32106311; 307922...   

                                         panel comments  \
0       Developmental disorders; Eye disorders      NaN   
1                      Developmental disorders      NaN   
2                      Developmental disorders      NaN   
3  Developmental disorders; Skeletal disorders      NaN   
4      Developmental disorders; Skin disorders      NaN   

        date_of_last_review                                     g2p_record_url  
0 2019-09-26 16:23:46+00:00  https://www.ebi.ac.uk/gene2phenotype/lgd/G2P00001  
1 2025-01-28 23:09:54+00:00  https://www.ebi.ac.uk/gene2phenotype/lgd/G2P00002  
2 2015-07-22 16:14:07+00:00  https://www.ebi.ac.uk/gene2phenotype/lgd/G2P00003  
3 2025-01-27 14:24:27+00:00  https://www.ebi.ac.uk/gene2phenotype/lgd/G2P00004  
4 2024-08-20 14:13:58+00:00  https://www.ebi.ac.uk/gene2phenotype/lgd/G2P00005

## Pre-NodeNorming

Querying NodeNorm: send unique values (no duplicates!) from entire column in large batches -> generate mapping dict to use. 
<br>
__Not querying 1-by-1 or 1 row at a time: much slower__ and would involve sending duplicate IDs (unless saved dict is kept outside loop and checked) 

Not going to use NameResolver: not optimistic this would work anyways. My manual process of getting "better" disease IDs is to use the gene IDs, find the diseases they're linked to in OMIM and Monarch, and seeing if those match the data's disease name / phenotypes / publications. This is more complicated than just using NameResolver.

<div class="alert alert-block alert-danger">

Set the NodeNorm URL you want to use. 

In [18]:
## put into parser (format): DONE

import requests

## from BioThings annotator code: for interoperability between diff Python versions
# try:
#     from itertools import batched  # new in Python 3.12
# except ImportError:
#     from itertools import islice

#     def batched(iterable, n):
#         # batched('ABCDEFG', 3) → ABC DEF G
#         if n < 1:
#             raise ValueError("n must be at least one")
#         iterator = iter(iterable)
#         while batch := tuple(islice(iterator, n)):
#             yield batch

## doing to test that this works
from itertools import islice

def batched(iterable, n):
    # batched('ABCDEFG', 3) → ABC DEF G
    if n < 1:
        raise ValueError("n must be at least one")
    iterator = iter(iterable)
    while batch := tuple(islice(iterator, n)):
        yield batch

nodenorm_url = "https://nodenorm.ci.transltr.io/get_normalized_nodes"

### Exploring: Genes

**2025-03-28 data:** Every row has at least 1 gene ID (HGNC column has no missing values). So no rows will be removed because there's no gene IDs to use for the pre-NodeNorming. 

In [19]:
df[["gene_symbol", "hgnc_id", "gene_mim"]].info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3659 entries, 0 to 3658
Data columns (total 3 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   gene_symbol  3659 non-null   object
 1   hgnc_id      3659 non-null   object
 2   gene_mim     3657 non-null   object
dtypes: object(3)
memory usage: 85.9+ KB


In [20]:
df[df["gene_mim"].isna()]

g2p_id gene_symbol gene_mim     hgnc_id previous_gene_symbols  \
955   G2P00977      ZNF599      NaN  HGNC:26408              FLJ30663   
3452  G2P02168      MFSD6L      NaN  HGNC:26656              FLJ35773   

                            disease_name disease_mim  disease_MONDO  \
955           ZNF599-related NOT IN OMIM         NaN            NaN   
3452  MFSD6L-related congenital cataract         NaN  MONDO:0005129   

        allelic_requirement cross_cutting_modifier confidence  \
955   monoallelic_autosomal                    NaN    limited   
3452    biallelic_autosomal                    NaN    limited   

                 variant_consequence  \
955                        uncertain   
3452  altered gene product structure   

                                          variant_types molecular_mechanism  \
955                                                 NaN        undetermined   
3452  inframe_insertion; inframe_deletion; missense_...        undetermined   

     molecular_mechanism_categorisation molecular_mechanism_evidence  \
955                            inferred                          NaN   
3452                           inferred                          NaN   

                  phenotypes publications                    panel comments  \
955                      NaN          NaN  Developmental disorders      NaN   
3452  HP:0000007; HP:0010864     22935719            Eye disorders      NaN   

           date_of_last_review  \
955  2015-07-22 16:15:03+00:00   
3452 2017-08-29 09:35:13+00:00   

                                         g2p_record_url  
955   https://www.ebi.ac.uk/gene2phenotype/lgd/G2P00977  
3452  https://www.ebi.ac.uk/gene2phenotype/lgd/G2P02168

In [21]:
## saving stats on data with no gene IDs, just in case

stats_no_gene_IDs = {
    "n_rows": df[df["gene_mim"].isna() & df["hgnc_id"].isna()].shape[0],
    "n_names": len(df[df["gene_mim"].isna() & df["hgnc_id"].isna()]["gene_symbol"].unique())
}

stats_no_gene_IDs["n_rows"]
stats_no_gene_IDs["n_names"]

0

0

#### HGNC

__Running Gene HGNC IDs through NodeNorm__


Catching potential mapping failures for later stats report

In [22]:
## saving stats on data with no HGNC IDs, just in case

n_rows_no_hgnc = df["hgnc_id"].isna().sum()

In [23]:
## get set of unique CURIEs to put into NodeNorm
hgnc = df["hgnc_id"].dropna().unique()
len(hgnc)

3000

In [24]:
hgnc_nodenorm_mapping = {}

## set up variables to catch potential mapping failures
stats_hgnc_mapping_failures = {
    "unexpected_error": {},
    "nodenorm_returned_none": [],
    "wrong_category": {},
    "no_label": []
    
}

In [25]:
## larger batches are quicker
for batch in batched(hgnc, 1000):
    ## returns tuples -> cast to list
    req_body = {
        "curies": list(batch),
        "conflate": True,
    }
    r = requests.post(nodenorm_url, json=req_body)
    response = r.json()
    
    ## not doing dict comprehension. allows easier review, logic writing
    for k,v in response.items():
        ## catch unexpected errors
        try:
            ## if NodeNorm didn't have info on this ID, v will be None
            if v is not None:
                ## don't keep mapping if category is not the expected one
                if v["type"][0] == "biolink:Gene":
                    ## also throw out mapping if no primary label found
                    if v["id"].get("label"):
                        temp = {
                            k: {"primary_id": v["id"]["identifier"],
                                "primary_label": v["id"]["label"]
                               }
                        }
                        hgnc_nodenorm_mapping.update(temp)
                    else:
                        stats_hgnc_mapping_failures["no_label"].append(k)
#                         print(f"{k}: NodeNorm didn't find primary label. Not keeping this mapping.")
                else:
                    stats_hgnc_mapping_failures["wrong_category"].update({k: v["type"][0]})
#                     print(f'{k}: NodeNorm found different category {v["type"][0]}. Not keeping this mapping.')
            else:
                stats_hgnc_mapping_failures["nodenorm_returned_none"].append(k)
#                 print(f"{k}: NodeNorm didn't recognize this ID")
        except:
            stats_hgnc_mapping_failures["unexpected_error"].update({k: v})
            print(f'Encountered an unexpected error.')
            print(f'NodeNorm response key: {k}')
            print(f'NodeNorm response value: {v}')

In [26]:
len(hgnc_nodenorm_mapping)

stats_hgnc_mapping_failures

3000

{'unexpected_error': {},
 'nodenorm_returned_none': [],
 'wrong_category': {},
 'no_label': []}

#### OMIM

__Running Gene OMIM IDs through NodeNorm__

Catching potential mapping failures for later stats report. 

Pasted, adjusted from HGNC code blocks above.

In [27]:
## get set of unique CURIEs to put into NodeNorm
gene_omim = df["gene_mim"].dropna().unique()
len(gene_omim)

2998

In [28]:
gene_omim_nodenorm_mapping = {}

## set up variables to catch potential mapping failures
stats_gene_omim_mapping_failures = {
    "unexpected_error": {},
    "nodenorm_returned_none": [],
    "wrong_category": {},
    "no_label": []
    
}

In [29]:
## larger batches are quicker
for batch in batched(gene_omim, 1000):
    ## returns tuples -> cast to list
    req_body = {
        "curies": list(batch),
        "conflate": True,
    }
    r = requests.post(nodenorm_url, json=req_body)
    response = r.json()
    
    ## not doing dict comprehension. allows easier review, logic writing
    for k,v in response.items():
        ## catch unexpected errors
        try:
            ## if NodeNorm didn't have info on this ID, v will be None
            if v is not None:
                ## don't keep mapping if category is not the expected one
                if v["type"][0] == "biolink:Gene":
                    ## also throw out mapping if no primary label found
                    if v["id"].get("label"):
                        temp = {
                            k: {"primary_id": v["id"]["identifier"],
                                "primary_label": v["id"]["label"]
                               }
                        }
                        gene_omim_nodenorm_mapping.update(temp)
                    else:
                        stats_gene_omim_mapping_failures["no_label"].append(k)
#                         print(f"{k}: NodeNorm didn't find primary label. Not keeping this mapping.")
                else:
                    stats_gene_omim_mapping_failures["wrong_category"].update({k: v["type"][0]})
#                     print(f'{k}: NodeNorm found different category {v["type"][0]}. Not keeping this mapping.')
            else:
                stats_gene_omim_mapping_failures["nodenorm_returned_none"].append(k)
#                 print(f"{k}: NodeNorm didn't recognize this ID")
        except:
            stats_gene_omim_mapping_failures["unexpected_error"].update({k: v})
            print(f'Encountered an unexpected error.')
            print(f'NodeNorm response key: {k}')
            print(f'NodeNorm response value: {v}')

In [30]:
len(gene_omim_nodenorm_mapping)

stats_gene_omim_mapping_failures

2997

{'unexpected_error': {},
 'nodenorm_returned_none': ['OMIM:621003'],
 'wrong_category': {},
 'no_label': []}

In [31]:
## from looking at 2025-03-28 data
df[df["gene_mim"] == "OMIM:621003"]

g2p_id gene_symbol     gene_mim     hgnc_id previous_gene_symbols  \
2766  G2P03714         SP9  OMIM:621003  HGNC:30690                ZNF990   

                                           disease_name disease_mim  \
2766  SP9-related neurodevelopmental disorder with/w...         NaN   

     disease_MONDO    allelic_requirement cross_cutting_modifier confidence  \
2766           NaN  monoallelic_autosomal                    NaN   moderate   

                 variant_consequence  \
2766  altered gene product structure   

                                          variant_types molecular_mechanism  \
2766  frameshift_variant_NMD_escaping; missense_variant        undetermined   

     molecular_mechanism_categorisation molecular_mechanism_evidence  \
2766                           inferred                          NaN   

     phenotypes publications                    panel  \
2766        NaN     38288683  Developmental disorders   

                                               comments  \
2766  Discussions during curation agreed that there ...   

           date_of_last_review  \
2766 2025-03-05 11:14:46+00:00   

                                         g2p_record_url  
2766  https://www.ebi.ac.uk/gene2phenotype/lgd/G2P03714

**2025-03-28 data:** `OMIM:621003` is a valid gene ID [(OMIM webpage)](https://omim.org/entry/602668), so this is a NodeNorm error.

#### Comparing HGNC vs OMIM

In [32]:
## if row has both IDs, look for diff in mappings from each ID
for row in df[["gene_mim", "hgnc_id"]].itertuples(index=False):
    ## has both IDs
    if pd.notna(row.gene_mim) and pd.notna(row.hgnc_id):
        ## if have NodeNorm mappings for both
        if gene_omim_nodenorm_mapping.get(row.gene_mim) and \
        hgnc_nodenorm_mapping.get(row.hgnc_id):
            ## check if mappings are diff
            if gene_omim_nodenorm_mapping[row.gene_mim]["primary_id"] != \
            hgnc_nodenorm_mapping[row.hgnc_id]["primary_id"]:
                print(row)

## 2025-03-28 data: nothing prints, so there are no mismatches

In [33]:
## look for differences in name between NodeNormed and original data

for row in df[["gene_symbol", "hgnc_id"]].itertuples(index=False):
    ## works because both columns have no missing values and there's no failed mappings
    ## if this changes, need to adjust this code block
    if row.gene_symbol != hgnc_nodenorm_mapping[row.hgnc_id]["primary_label"]:
        print(f"G2P name {row.gene_symbol}, ID {row.hgnc_id}")
        print(f'NodeNorm name {hgnc_nodenorm_mapping[row.hgnc_id]["primary_label"]}, ID {hgnc_nodenorm_mapping[row.hgnc_id]["primary_id"]}')
        print("\n")

G2P name MT-TP, ID HGNC:7494
NodeNorm name TRNP, ID NCBIGene:4571


G2P name CENPJ, ID HGNC:17272
NodeNorm name CPAP, ID NCBIGene:55835


G2P name CCDC103, ID HGNC:32700
NodeNorm name DNAAF19, ID NCBIGene:388389


G2P name MT-TL1, ID HGNC:7490
NodeNorm name TRNL1, ID NCBIGene:4567


G2P name MT-ND1, ID HGNC:7455
NodeNorm name ND1, ID NCBIGene:4535


G2P name MT-ND4, ID HGNC:7459
NodeNorm name ND4, ID NCBIGene:4538


G2P name MT-ATP6, ID HGNC:7414
NodeNorm name ATP6, ID NCBIGene:4508


G2P name MT-ND5, ID HGNC:7461
NodeNorm name ND5, ID NCBIGene:4540


G2P name MT-ND6, ID HGNC:7462
NodeNorm name ND6, ID NCBIGene:4541




**2025-03-28 data:** 

Review of mismatched names:
* NodeNorm is correct that CENPJ should be CPAP, CCDC103 -> DNAAF19
* The rest look like mitochondrial genes, and NCBIGene main name seems to match G2P name, not NodeNorm -> messaged NodeNorm

#### Conclusions

<div class="alert alert-block alert-success">

**2025-03-28 data:** 
    
__Exploration__

* no mapping failures
* when rows have both OMIM and HGNC IDs, there were no differences in NodeNorm mapping ("mismatches")
    
__Decision: Use HGNC ID column to generate NodeNorm values__

* less missing values (none right now)
* these IDs are probably only genes (vs OMIM terms can be multiple types) 

### Exploring: Diseases

There are many more missing IDs for Disease, compared to Gene. 

As mentioned at the beginning of the "Pre-NodeNorming" section, I won't be using NameResolver right now. 

__This means all rows w/o any disease IDs will be removed__ because they cannot be pre-NodeNormed. 

In [34]:
df[["disease_name", "disease_mim", "disease_MONDO"]].info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3659 entries, 0 to 3658
Data columns (total 3 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   disease_name   3659 non-null   object
 1   disease_mim    2572 non-null   object
 2   disease_MONDO  564 non-null    object
dtypes: object(3)
memory usage: 85.9+ KB


In [35]:
df[df["disease_mim"].isna() & df["disease_MONDO"].isna()]

g2p_id gene_symbol     gene_mim     hgnc_id  \
15    G2P00016       NAA10  OMIM:300013  HGNC:18704   
20    G2P00021      COL1A1  OMIM:120150   HGNC:2197   
32    G2P00033       SCN2A  OMIM:182390  HGNC:10588   
56    G2P00057        FLNA  OMIM:300017   HGNC:3754   
83    G2P00086       PRRT2  OMIM:614386  HGNC:30500   
...        ...         ...          ...         ...   
3621  G2P02428       IPO13  OMIM:610411  HGNC:16853   
3624  G2P02432        MAFB  OMIM:608968   HGNC:6408   
3630  G2P02540       IFT88  OMIM:600595  HGNC:20606   
3646  G2P02823        MFRP  OMIM:606227  HGNC:18121   
3649  G2P02826       RPE65  OMIM:180069  HGNC:10294   

                                  previous_gene_symbols  \
15                             ARD1; ARD1A; DXS707; TE2   
20                                                  OI4   
32                HBSCI; HBSCII; NAV1.2; SCN2A1; SCN2A2   
56                       ABP-280; FLN; FLN1; OPD1; OPD2   
83    DKFZP547J199; DSPB3; DYT10; EKD1; FICCA; FLJ25...   
...                                                 ...   
3621                           IMP13; KIAA0724; RANBP13   
3624                                               KRML   
3630          D13S1056E; HTG737; MGC26259; TG737; TTC10   
3646                       C1QTNF5; FLJ30570; NNO2; RD6   
3649                             BCO3; LCA2; RD12; RP20   

                                           disease_name disease_mim  \
15    NAA10-related nonpecific severe intellectual d...         NaN   
20      COL1A1-related osteogenesis imperfecta spectrum         NaN   
32    SCN2A-related nonspecific severe intellectual ...         NaN   
56                FLNA-related epileptic encephalopathy         NaN   
83    PRRT2-related intellectual developmental disorder         NaN   
...                                                 ...         ...   
3621  IPO13-related ocular coloboma, microphthalmia,...         NaN   
3624  MAFB-related focal segmental glomerulosclerosi...         NaN   
3630   IFT88-related non-syndromic retinal degeneration         NaN   
3646      MFRP-related non-syndromic retinitis pigmenta         NaN   
3649                    RPE65-related retinal dystrophy         NaN   

     disease_MONDO         allelic_requirement   cross_cutting_modifier  \
15             NaN  monoallelic_X_heterozygous                      NaN   
20             NaN       monoallelic_autosomal  restricted mutation set   
32             NaN       monoallelic_autosomal                      NaN   
56             NaN  monoallelic_X_heterozygous                      NaN   
83             NaN         biallelic_autosomal                      NaN   
...            ...                         ...                      ...   
3621           NaN         biallelic_autosomal                      NaN   
3624           NaN       monoallelic_autosomal                      NaN   
3630           NaN         biallelic_autosomal                      NaN   
3646           NaN         biallelic_autosomal                      NaN   
3649           NaN       monoallelic_autosomal                      NaN   

      confidence             variant_consequence  \
15    definitive  altered gene product structure   
20    definitive  altered gene product structure   
32    definitive             absent gene product   
56    definitive  altered gene product structure   
83    definitive             absent gene product   
...          ...                             ...   
3621      strong             absent gene product   
3624      strong  altered gene product structure   
3630     limited             absent gene product   
3646  definitive                       uncertain   
3649  definitive  altered gene product structure   

                                          variant_types molecular_mechanism  \
15    inframe_insertion; inframe_deletion; missense_...        undetermined   
20                                                  NaN   dominant negative   
32                    

In [36]:
## saving stats on data with no disease IDs

stats_no_disease_IDs = {
    "n_rows": df[df["disease_mim"].isna() & df["disease_MONDO"].isna()].shape[0],
    "n_names": len(df[df["disease_mim"].isna() & df["disease_MONDO"].isna()]["disease_name"].unique())
}

stats_no_disease_IDs["n_rows"]
stats_no_disease_IDs["n_names"]

642

629

#### OMIM/orphanet

__Running OMIM/orphanet IDs through NodeNorm__

Catching mapping failures for later stats report

Pasted, adjusted from HGNC code blocks above.

In [37]:
## put into parser (format): DONE

## get set of unique CURIEs to put into NodeNorm
disease_OmOr = df["disease_mim"].dropna().unique()
len(disease_OmOr)

2403

In [38]:
## put into parser (format): DONE

OmOr_nodenorm_mapping = {}

## set up variables to catch mapping failures
stats_OmOr_mapping_failures = {
    "unexpected_error": {},
    "nodenorm_returned_none": [],
    "wrong_category": {},
    "no_label": []
    
}

In [39]:
## put into parser (format): DONE

## larger batches are quicker
for batch in batched(disease_OmOr, 1000):
    ## returns tuples -> cast to list
    req_body = {
        "curies": list(batch),
        "conflate": True,
    }
    r = requests.post(nodenorm_url, json=req_body)
    response = r.json()
    
    ## not doing dict comprehension. allows easier review, logic writing
    for k,v in response.items():
        ## catch unexpected errors
        try:
            ## if NodeNorm didn't have info on this ID, v will be None
            if v is not None:
                ## don't keep mapping if category is not the expected one
                if v["type"][0] == "biolink:Disease":
                    ## also throw out mapping if no primary label found
                    if v["id"].get("label"):
                        temp = {
                            k: {"primary_id": v["id"]["identifier"],
                                "primary_label": v["id"]["label"]
                               }
                        }
                        OmOr_nodenorm_mapping.update(temp)
                    else:
                        stats_OmOr_mapping_failures["no_label"].append(k)
#                         print(f"{k}: NodeNorm didn't find primary label. Not keeping this mapping.")
                else:
                    stats_OmOr_mapping_failures["wrong_category"].update({k: v["type"][0]})
#                     print(f'{k}: NodeNorm found different category {v["type"][0]}. Not keeping this mapping.')
            else:
                stats_OmOr_mapping_failures["nodenorm_returned_none"].append(k)
#                 print(f"{k}: NodeNorm didn't recognize this ID")
        except:
            stats_OmOr_mapping_failures["unexpected_error"].update({k: v})
            print(f'Encountered an unexpected error.')
            print(f'NodeNorm response key: {k}')
            print(f'NodeNorm response value: {v}')

In [40]:
## put into parser (format): DONE

## calculate stats: number of rows affected by each type of mapping failure
n_rows_no_data = df[
    df["disease_mim"].isin(stats_OmOr_mapping_failures["nodenorm_returned_none"])
].shape[0]
n_rows_wrong_category = df[
    df["disease_mim"].isin(stats_OmOr_mapping_failures["wrong_category"].keys())
].shape[0]
n_rows_no_label = df[
    df["disease_mim"].isin(stats_OmOr_mapping_failures["no_label"])
].shape[0]

In [41]:
len(OmOr_nodenorm_mapping)

stats_OmOr_mapping_failures["unexpected_error"]

len(stats_OmOr_mapping_failures["nodenorm_returned_none"])
len(stats_OmOr_mapping_failures["wrong_category"])
len(stats_OmOr_mapping_failures["no_label"])

2335

{}

39

26

3

In [42]:
## code used to review mapping failures 

# stats_OmOr_mapping_failures["nodenorm_returned_none"]

stats_OmOr_mapping_failures["wrong_category"].keys()

# stats_OmOr_mapping_failures["no_label"]

dict_keys(['OMIM:300153', 'OMIM:300171', 'OMIM:188400', 'OMIM:611579', 'OMIM:603707', 'OMIM:601789', 'OMIM:603360', 'OMIM:609413', 'OMIM:300197', 'OMIM:603164', 'OMIM:123580', 'OMIM:601758', 'OMIM:601498', 'OMIM:601791', 'OMIM:170993', 'OMIM:601893', 'OMIM:606525', 'OMIM:602859', 'OMIM:612082', 'OMIM:601757', 'OMIM:300204', 'OMIM:600112', 'OMIM:104155', 'OMIM:618615', 'OMIM:614281', 'OMIM:603718'])

In [43]:
## code used to review mapping failures 

# df[df["disease_mim"] == "OMIM:613180"]

<div class="alert alert-block alert-info">

**2025-02-28 data:**    
    
__Reviewed Disease OMIM/orphanet NodeNorm mapping failures__

All were OMIM IDs, none were orphanet. 
    
39 cases where NodeNorm returned None (didn't recognize/resolve ID). __I checked some (10)__:
* 5: ID has been replaced/moved to a diff ID (OMIM:607236, OMIM:608890, OMIM:613180, OMIM:300706, OMIM:300141) -> emailed EBI gene2pheno
* 3: ID doesn't exist (OMIM:249163, OMIM:319029, OMIM:237145) -> emailed EBI gene2pheno
* 1: NodeNorm error - this is a valid disease ID that it should recognize (OMIM:133700) -> messaged NodeNorm
* 1: valid ID, but it doesn't seem to be a disease. There may be better IDs out there (OMIM:601884) -> messaged NodeNorm, emailed EBI gene2pheno

26 cases where NodeNorm category was something else (currently, always Gene). I checked all: 
* 25: NodeNorm is correct, this is a gene -> emailed EBI gene2pheno
* 1: NodeNorm error - this is a valid disease ID (OMIM:188400) -> messaged NodeNorm
    
3 cases where NodeNorm didn't have a primary label. I checked all:
* 2: NodeNorm error - these are valid disease IDs with labels (OMIM:620987, OMIM:620964) -> messaged NodeNorm
* 1: valid ID, but it doesn't seem to be a disease (OMIM:300129). EBI gene2pheno shouldn't use, not sure it should be in NodeNorm -> messaged NodeNorm, emailed EBI gene2pheno
    
</div>

In [44]:
## code used to check for orphanet mapping failures 

for i in stats_OmOr_mapping_failures["nodenorm_returned_none"]:
    if "orphanet" in i:
        print(i)
        
for i in stats_OmOr_mapping_failures["wrong_category"].keys():
    if "orphanet" in i:
        print(i)

for i in stats_OmOr_mapping_failures["no_label"]:
    if "orphanet" in i:
        print(i)

**2025-02-28 data:** 

No orphanet IDs were had mapping failures but I checked all (9) mappings anyways - they looked fine. 

In [45]:
df[df["disease_mim"].str.contains("orphanet", na=False)]

g2p_id gene_symbol     gene_mim     hgnc_id  \
1959  G2P02555       SMAD6  OMIM:602931   HGNC:6772   
1968  G2P02564       TONSL  OMIM:604546   HGNC:7801   
2792  G2P01819         RET  OMIM:164761   HGNC:9967   
2913  G2P02669        INSR  OMIM:147670   HGNC:6091   
3143  G2P03354      SEC23B  OMIM:610512  HGNC:10702   
3316  G2P02001      RCBTB1  OMIM:607867  HGNC:18243   
3496  G2P02222       REV3L  OMIM:602776   HGNC:9968   
3538  G2P02286      TMEM98  OMIM:615949  HGNC:24529   
3619  G2P02426       TENM3  OMIM:610083  HGNC:29944   

                                  previous_gene_symbols  \
1959                             HST17432; MADH6; MADH7   
1968                                      IKBR; NFKBIL2   
2792  CDHF12; CDHR16; HSCR1; MEN2A; MEN2B; MTC1; PTC...   
2913                                              CD220   
3143                       CDA-II; CDAII; CDAN2; HEMPAS   
3316                             CLLD7; CLLL7; FLJ10716   
3496                                         POLZ; REV3   
3538                                      DKFZP564K1964   
3619                       KIAA1455; ODZ3; TEN-M3; TEN3   

                                           disease_name      disease_mim  \
1959       SMAD6-related non-syndromic craniosynostosis  orphanet:139390   
1968                TONSL-related sponastrime dysplasia   orphanet:93357   
2792            RET-related medullary thyroid carcinoma    orphanet:1332   
2913                         INSR-related leprechaunism     orphanet:508   
3143                     SEC23B-related Cowden syndrome     orphanet:201   
3316  RCBTB1-related familial exudative vitreoretino...     orphanet:891   
3496                     REV3L-related Moebius syndrome     orphanet:570   
3538                       TMEM98-related nanophthalmos   orphanet:35612   
3619          TENM3-related colobomatous microphthalmia   orphanet:98938   

     disease_MONDO    allelic_requirement cross_cutting_modifier  confidence  \
1959           NaN  monoallelic_autosomal                    NaN     limited   
1968           NaN    biallelic_autosomal                    NaN      strong   
2792           NaN  monoallelic_autosomal                    NaN  definitive   
2913           NaN    biallelic_autosomal                    NaN  definitive   
3143           NaN  monoallelic_autosomal                    NaN     limited   
3316           NaN  monoallelic_autosomal                    NaN     limited   
3496           NaN  monoallelic_autosomal                    NaN     limited   
3538           NaN  monoallelic_autosomal                    NaN      strong   
3619           NaN    biallelic_autosomal                    NaN      strong   

                 variant_consequence  \
1959             absent gene product   
1968             absent gene product   
2792                       uncertain   
2913             absent gene product   
3143                       uncertain   
3316  altered gene product structure   
3496                       uncertain   
3538                       uncertain   
3619             absent gene product   

                                          variant_types molecular_mechanism  \
1959                                                NaN    loss of function   
1968                                                NaN    loss of function   
2792                                                NaN        undetermined   
2913                                                NaN    loss of function   
3143                                                NaN    gain of function   
3316  inframe_insertion; inframe_deletion; missense_...        undetermined   
3496                                                NaN        undetermined   
3538                                                NaN        undetermined   
3619                                                NaN    loss of function   

     molecular_mechanism_categorisation molecular_mechanism_evidence  \
1959                           inferred 

In [46]:
OmOr_nodenorm_mapping["orphanet:1332"]

{'primary_id': 'MONDO:0015277',
 'primary_label': 'medullary thyroid gland carcinoma'}

<div class="alert alert-block alert-success">

**2025-03-28 data:** 
    
I decided <b>not to try using MONDO mappings when the OMIM mapping failed</b>, because there's only a few cases where those rows even have MONDO IDs to use. 

* nodenorm_returned_none (39): none have MONDO
* wrong_category (26): only 2 have MONDO
* no_label (3): none have MONDO

In [47]:
## code used to check how many rows have OMIM failure + MONDO ID 

df[df["disease_mim"].isin(stats_OmOr_mapping_failures["nodenorm_returned_none"]) & 
   df["disease_MONDO"].notna()].shape

df[df["disease_mim"].isin(stats_OmOr_mapping_failures["wrong_category"].keys()) & 
   df["disease_MONDO"].notna()].shape

df[df["disease_mim"].isin(stats_OmOr_mapping_failures["no_label"]) & 
   df["disease_MONDO"].notna()].shape

(0, 22)

(2, 22)

(0, 22)

#### MONDO

__Running MONDO IDs through NodeNorm__

Catching potential mapping failures for later stats report

Pasted, adjusted from Disease OMIM/orphanet code blocks above.

In [48]:
## get set of unique CURIEs to put into NodeNorm
mondo = df["disease_MONDO"].dropna().unique()
len(mondo)

383

In [49]:
mondo_nodenorm_mapping = {}

## set up variables to catch mapping failures
stats_mondo_mapping_failures = {
    "unexpected_error": {},
    "nodenorm_returned_none": [],
    "wrong_category": {},
    "no_label": []
    
}

In [50]:
## larger batches are quicker
for batch in batched(mondo, 1000):
    ## returns tuples -> cast to list
    req_body = {
        "curies": list(batch),
        "conflate": True,
    }
    r = requests.post(nodenorm_url, json=req_body)
    response = r.json()
    
    ## not doing dict comprehension. allows easier review, logic writing
    for k,v in response.items():
        ## catch unexpected errors
        try:
            ## if NodeNorm didn't have info on this ID, v will be None
            if v is not None:
                ## don't keep mapping if category is not the expected one
                if v["type"][0] == "biolink:Disease":
                    ## also throw out mapping if no primary label found
                    if v["id"].get("label"):
                        temp = {
                            k: {"primary_id": v["id"]["identifier"],
                                "primary_label": v["id"]["label"]
                               }
                        }
                        mondo_nodenorm_mapping.update(temp)
                    else:
                        stats_mondo_mapping_failures["no_label"].append(k)
#                         print(f"{k}: NodeNorm didn't find primary label. Not keeping this mapping.")
                else:
                    stats_mondo_mapping_failures["wrong_category"].update({k: v["type"][0]})
#                     print(f'{k}: NodeNorm found different category {v["type"][0]}. Not keeping this mapping.')
            else:
                stats_mondo_mapping_failures["nodenorm_returned_none"].append(k)
#                 print(f"{k}: NodeNorm didn't recognize this ID")
        except:
            stats_mondo_mapping_failures["unexpected_error"].update({k: v})
            print(f'Encountered an unexpected error.')
            print(f'NodeNorm response key: {k}')
            print(f'NodeNorm response value: {v}')

In [51]:
len(mondo_nodenorm_mapping)

stats_mondo_mapping_failures

383

{'unexpected_error': {},
 'nodenorm_returned_none': [],
 'wrong_category': {},
 'no_label': []}

#### Comparing OMIM/orphanet vs MONDO

In [52]:
## if row has both IDs, look for diff in mappings from each ID

## list of tuples (omim/orpha, mondo)
mismatches = []

for row in df[["disease_mim", "disease_MONDO"]].itertuples(index=False):
    ## has both IDs
    if pd.notna(row.disease_mim) and pd.notna(row.disease_MONDO):
        ## if have NodeNorm mappings for both
        if OmOr_nodenorm_mapping.get(row.disease_mim) and \
        mondo_nodenorm_mapping.get(row.disease_MONDO):
            ## check if mappings are diff
            if OmOr_nodenorm_mapping[row.disease_mim]["primary_id"] != \
            mondo_nodenorm_mapping[row.disease_MONDO]["primary_id"]:
                mismatches.append((row.disease_mim, row.disease_MONDO))

print(f"There's {len(mismatches)} mismatches between the OMIM/orphanet and MONDO NodeNorm mappings.")

There's 22 mismatches between the OMIM/orphanet and MONDO NodeNorm mappings.


In [53]:
## code chunk to review mismatches 1 by 1
mismatches[21]

('OMIM:175800', 'MONDO:0006602')

In [54]:
## code chunk to review mismatches 1 by 1

OmOr_nodenorm_mapping["OMIM:175800"]
mondo_nodenorm_mapping["MONDO:0006602"]

{'primary_id': 'MONDO:0008290',
 'primary_label': 'porokeratosis 1, Mibelli type'}

{'primary_id': 'MONDO:0006602', 'primary_label': 'porokeratosis'}

In [55]:
## code chunk to review mismatches 1 by 1

df[df["disease_mim"] == "OMIM:300696"]
df[df["disease_MONDO"] == "MONDO:0013812"]

g2p_id gene_symbol     gene_mim    hgnc_id  \
2524  G2P03293        FHL1  OMIM:300163  HGNC:3702   

                                  previous_gene_symbols  \
2524  BA535K18.1; FHL1B; FLH1A; KYO-T; MGC111107; SL...   

                                        disease_name  disease_mim  \
2524  FHL1-related Emery-Dreifuss muscular dystrophy  OMIM:300696   

      disease_MONDO       allelic_requirement cross_cutting_modifier  \
2524  MONDO:0010680  monoallelic_X_hemizygous                    NaN   

      confidence                                variant_consequence  \
2524  definitive  decreased gene product level; absent gene prod...   

                                          variant_types molecular_mechanism  \
2524  stop_gained_NMD_escaping; stop_lost; splice_re...    loss of function   

     molecular_mechanism_categorisation molecular_mechanism_evidence  \
2524                           inferred                          NaN   

                                             phenotypes  \
2524  HP:0003306; HP:0003704; HP:0003805; HP:0003676...   

                                           publications  \
2524  18179888; 19687455; 30681346; 19716112; 201868...   

                                           panel  \
2524  Developmental disorders; Cardiac disorders   

                                               comments  \
2524  Expert review done on 12/01/2022; FHL1-related...   

           date_of_last_review  \
2524 2024-03-26 10:33:21+00:00   

                                         g2p_record_url  
2524  https://www.ebi.ac.uk/gene2phenotype/lgd/G2P03293

g2p_id gene_symbol     gene_mim   hgnc_id previous_gene_symbols  \
122   G2P00126        ACTB  OMIM:102630  HGNC:132                   NaN   
1610  G2P01638       ACTG1  OMIM:102560  HGNC:144  ACTG; DFNA20; DFNA26   

                                 disease_name  disease_mim  disease_MONDO  \
122    ACTB-related Baraitser-Winter syndrome  OMIM:243310  MONDO:0013812   
1610  ACTG1-related Baraitser-Winter syndrome  OMIM:614583  MONDO:0013812   

        allelic_requirement   cross_cutting_modifier  confidence  \
122   monoallelic_autosomal  restricted mutation set  definitive   
1610  monoallelic_autosomal  restricted mutation set  definitive   

                 variant_consequence  \
122   altered gene product structure   
1610  altered gene product structure   

                                          variant_types molecular_mechanism  \
122   inframe_insertion; inframe_deletion; missense_...    gain of function   
1610  inframe_insertion; inframe_deletion; missense_...    gain of function   

     molecular_mechanism_categorisation molecular_mechanism_evidence  \
122                            inferred                          NaN   
1610                           inferred                          NaN   

                                             phenotypes  \
122   HP:0001274; HP:0002162; HP:0008897; HP:0005487...   
1610  HP:0000159; HP:0001167; HP:0001274; HP:0000465...   

                                          publications  \
122   22366783; 25052316; 27625340; 38592426; 34970860   
1610  22366783; 25052316; 27096712; 27240540; 29024830   

                                       panel comments  \
122   Developmental disorders; Eye disorders      NaN   
1610  Developmental disorders; Eye disorders      NaN   

           date_of_last_review  \
122  2015-07-22 16:14:17+00:00   
1610 2015-07-22 16:15:41+00:00   

                                         g2p_record_url  
122   https://www.ebi.ac.uk/gene2phenotype/lgd/G2P00126  
1610  https://www.ebi.ac.uk/gene2phenotype/lgd/G2P01638

<div class="alert alert-block alert-info">    

**2025-02-28 data:** 

__Review of OMIM vs MONDO NodeNorm mismatches (22)__

None were orphanet.    
    
---

__19: OMIM's mapping is better__

> __6: Mondo ID is related but wrong__ -> emailed EBI gene2pheno w/ example
> * 'OMIM:243310', 'MONDO:0013812': omim is correct syndrome 1, but mondo is syndrome 2 (diff gene)
> * 'OMIM:613575', 'MONDO:0044314': omim is correct 55, but mondo is 78 (diff gene)
> * 'OMIM:101000', 'MONDO:0008075': omim is correct type of schwannomatosis (NF2/type 2), vs mondo is a sibling. 
>   * NodeNorm should map omim to MONDO:0007039 but isn't -> messaged NodeNorm
> * 'OMIM:613987', __'MONDO:0009136'__: omim is correct recessive 2, but mondo is recessive 1 (diff gene? Confusing because Monarch page links to gene NHP2 but OMIM page doesn't)
>   * NodeNorm should map omim to MONDO:0013519 but isn't -> messaged NodeNorm  
> * 'OMIM:613988', 'MONDO:0009136': omim is correct recessive 3, but mondo is recessive 1 (diff gene)
>   * NodeNorm should map omim to MONDO:0013520 but isn't -> messaged NodeNorm
> * 'OMIM:616353', 'MONDO:0009136': omim is correct recessive 6, but mondo is recessive 1 (diff gene)
>   * NodeNorm should map omim to MONDO:0014600 but isn't -> messaged NodeNorm

> __13: Mondo ID is too general__ (can see on Monarch website) -> emailed EBI gene2pheno w/ example
> * 'OMIM:300696', 'MONDO:0010680': omim maps to MONDO:0010401, child of the mondo
> * 'OMIM:304120', 'MONDO:0019027': omim maps to MONDO:0010571 (syndrome type 2), child of the mondo (syndrome)
> * 'OMIM:610019', 'MONDO:0005129': omim maps to MONDO:0012395 (cataract 18), child of the mondo (cataract)
> * 'OMIM:611726', 'MONDO:0016295': omim maps to MONDO:0012721, child of the mondo 
> * 'OMIM:602668', 'MONDO:0016107': omim maps to MONDO:0011266 (type 2), child of the mondo
> * 'OMIM:203200', 'MONDO:0018910': omim maps to MONDO:0008746 (type 2), child of the mondo
> * 'OMIM:614328', 'MONDO:0017411': omim maps to MONDO:0013693 (type 1), child of the mondo
> * 'OMIM:175800', 'MONDO:0006602': omim maps to MONDO:0008290 (1, mibelli type), grandchild of the mondo
> * 'OMIM:614073', **'MONDO:0019312'**: omim maps to MONDO:0013556 (syndrome 4), child of the mondo (syndrome)
> * 'OMIM:614074', 'MONDO:0019312': omim maps to MONDO:0013557 (syndrome 5), child of the mondo (syndrome)
> * 'OMIM:614075', 'MONDO:0019312': omim maps to MONDO:0013558 (syndrome 6), child of the mondo (syndrome)
> * 'OMIM:614076', 'MONDO:0019312': omim maps to MONDO:0013559 (syndrome 7), child of the mondo (syndrome)
> * 'OMIM:614077', 'MONDO:0019312': omim maps to MONDO:0013560 (syndrome 8), child of the mondo (syndrome)

    
**1: MONDO's mapping is better**
<br>
Omim ID is slightly off -> __TELL EBI GENE2PHENO?__
* 'OMIM:613723', 'MONDO:0009181': mondo matches the disease name and phenotypes listed in the record better than the omim 


**1: Unsure**
* 'OMIM:158350', 'MONDO:0017623': omim is for Cowden syndrome 1, mondo is for PTEN hamartoma tumor syndrome. These are very similar, so I'm not sure which one is better. -> __TELL EBI GENE2PHENO?__
  * There's also another record w/ just the OMIM ID. I think the two rows should be merged. -> __TELL EBI GENE2PHENO?__


**1: NodeNorm error** -> messaged NodeNorm
* 'OMIM:224230', 'MONDO:0009136': both are recessive 1, NodeNorm should map to same entity

**Other rows reviewed:**
* 'OMIM:614583', 'MONDO:0013812': map to same correct entity

**2025-02-28 data:** 

The prelim decision is to use disease OMIM/orphanet IDs because:
* less missing values
* more accurate in cases where there's also a MONDO ID

#### Checking MONDO data

Above, I decided the OMIM/orphanet disease IDs were better. 

However, I wondered if the MONDO IDs were accurate to the disease name when there weren't OMIM/orphanet IDs. Then they could be used for NodeNorming and less data would be dropped because it wasn't pre-NodeNormed. 

In [56]:
## get the data that has MONDO, doesn't have OMIM/orphanet

df_mondo_only = df[df["disease_mim"].isna() & df["disease_MONDO"].notna()].copy()

mondo_only = df_mondo_only["disease_MONDO"].dropna().unique()

In [57]:
## saving stats on data with only MONDO ID

stats_mondo_only = {
    "n_rows": df_mondo_only.shape[0],
    "n_names": len(mondo_only)
}

stats_mondo_only["n_rows"]
stats_mondo_only["n_names"]

445

278

In [58]:
## code chunk used to review some of the data

# df_mondo_only[df_mondo_only["disease_MONDO"] == mondo_only[240]]

df_mondo_only[df_mondo_only["panel"].str.contains("Skeletal", na=False)]

g2p_id gene_symbol     gene_mim     hgnc_id previous_gene_symbols  \
3241  G2P03497        POP1  OMIM:602486  HGNC:30129                   NaN   

                          disease_name disease_mim  disease_MONDO  \
3241  POP1-related anauxetic dysplasia         NaN  MONDO:0011773   

      allelic_requirement cross_cutting_modifier  confidence  \
3241  biallelic_autosomal                    NaN  definitive   

               variant_consequence variant_types molecular_mechanism  \
3241  decreased gene product level           NaN        undetermined   

     molecular_mechanism_categorisation molecular_mechanism_evidence  \
3241                           inferred                          NaN   

     phenotypes                  publications               panel comments  \
3241        NaN  21455487; 27380734; 28067412  Skeletal disorders      NaN   

           date_of_last_review  \
3241 2023-12-20 09:04:04+00:00   

                                         g2p_record_url  
3241  https://www.ebi.ac.uk/gene2phenotype/lgd/G2P03497

<div class="alert alert-block alert-info">    

**2025-02-28 data:** 

__Reviewed some of the data__

Method: look at individual MONDO IDs. Covered all panels (only 1 skeletal, no hearing). 3 from earlier review (related to mismatches) + idx 0-240, step 10 + skeletal. 
    
__Summary__
* 37 rows (29 unique MONDO)
* __~16%__ were wrong (6/37) 
* Could tell EBI gene2pheno of issues but they are similar to those listed in mismatch mapping section

__Details__

__6 MONDO is related but wrong__
* "MONDO:0009136" for "RTEL1-related dyskeratosis congenita" (two rows): mondo is recessive 1, which is wrong. Should be recessive 5 MONDO:0014076/OMIM:615190 (old/synonym name is dominant 4) 
* "MONDO:0044314" for 4 rows "CLN3-related retinal dystrophy", "GUCA1B-", "PRPS1-", "SNRNP200-": mondo is type 78 (specifically for ARHGEF18), which is wrong. Should instead be:
  * CLN3 and PRPS1: a more general term like MONDO:0004580 (retinal degeneration) -> MONDO:0019118 (inherited retinal dystrophy) -> MONDO:0019200 (retinitis pigmentosa)
  * GUCA1B: type 48, MONDO:0013447
  * SNRNP200: type 33, MONDO:0012477
* "MONDO:0013522" for "TERC-related dyskeratosis congenita": mondo is for type 3 (specifically for TINF2, that row is in "Great" section). Should be type 1 MONDO:0007485/OMIM:127550. (confusing because Monarch's page of type 1 includes TINF2 and TERT too, but OMIM page only includes TERC)


__4 MONDO is too general__ 
* "MONDO:0020341" (periventricular nodular heterotopia) for "ERMARD-related periventricular heterotopia". The ERMARD-specific version is a child term: MONDO:0014240/OMIM:615544 (type 6)
* "MONDO:0018965" (Alport syndrome) for "COL4A5-related Alport syndrome". The COL4A5-specific version is a child term: MONDO:0010520/OMIM:301050  (X-linked)
* "MONDO:0024676" for "REST-related Wilms tumour": The REST-specific version is a **related** term: MONDO:0014779/OMIM:616806 (type 6)
* "MONDO:0011773" for "POP1-related anauxetic dysplasia": the POP1-specific version is a child term: MONDO:0054561/OMIM:617396 (type 2)


__4 Unsure -> TELL EBI GENE2PHENO?__
* "MONDO:0005129" for "CYP51A1-related congenital cataract": mondo is cataract, which is not wrong but kinda general. MONDO:0033853 seems better (correlated with gene, matches phenotypes, orphanet ref uses one of the ref papers) 
* "MONDO:0018869" for "TMTC3-related cobblestone lissencephaly": while the mondo (cobblestone lissencephaly) sounds correct, it isn't linked to this gene. VS another sibling disease is linked to the gene, matches phenotypes, uses same paper: MONDO:0014992/OMIM:617255 (lissencephaly 8)
* "MONDO:0100100" for "SELENON-related myopathy": while mondo has exact name match, it's not directly linked to gene. Instead, its child disease is directly linked to gene MONDO:0011271/OMIM:602771 (rigid spine muscular dystrophy 1)
* "MONDO:0020367" for "MYOC-related juvenile open angle glaucoma": while mondo is almost-exact name match, it's not directly linked to gene. Instead, its child disease is directly linked to gene MONDO:0007664/OMIM:137750 (glaucoma 1, open angle, A) 


__5 Okay (using general term is fine)__
* "MONDO:0005129" for 3 other rows "WDR87-related congenital cataract", "AKR1E2-", "MFSD6L-": couldn't find better mappings. 
* "MONDO:0015469" for "DHRS3 related craniosynostosis": couldn't find better mapping
* "MONDO:0024676" (childhood kidney Wilms tumor) for "CTR9-related Wilms tumour", "TRIM28-": couldn't find better mapping. TRIM28 is correlated to parent term (kidney Wilms tumor). 


__18 Great__
* "MONDO:0012506" for "DSC2-related arrhythmogenic right ventricular cardiomyopathy"
* "MONDO:0011001" for "SCN5A-related Brugada syndrome"
* "MONDO:0013262" for "MYH7-related dilated cardiomyopathy"
* "MONDO:0013369" for "TNNI3-related hypertrophic cardiomyopathy"
* "MONDO:0010946" for "PRKAG2-related cardiomyopathy"
* "MONDO:0014143" for "RIT1-related Noonan syndrome"
* "MONDO:0010015" for "PXDN-related anterior segment dysgenesis with sclerocornea"
* "MONDO:0014214" for "DYNC2I1-related short-rib polydactyly"
* "MONDO:0013522" for "TINF2-related dyskeratosis congenita"
* "MONDO:0032876" for "WASF1-related intellectual disability with seizures"
* "MONDO:0859164" for "UNC45A-related osteootohepatoenteric syndrome"
* "MONDO:0018772" for "SLC30A7-related Joubert syndrome": using general term is fine since there isn't any established subtype of Joubert syndrome for this gene
* "MONDO:0010215" for "ERCC4-related xeroderma pigmentosum, group F"
* "MONDO:0009735" for "SPINK5-related Netherton syndrome"
* "MONDO:0007808" for "KRT1-related ichthyosis hystrix, Curth-Macklin type"
* "MONDO:0007566" for "TGFBR1-related multiple self-healing squamous epithelioma"
* "MONDO:0008285" for "PDGFRA-related gastrointestinal stromal tumor/GIST-plus syndrome, somatic or familial"
* "MONDO:0010912" for "TUBB3-related fibrosis of extraocular muscles, congenital"

#### Conclusions

<div class="alert alert-block alert-success">

**2025-02-28 data:** 
    
__Exploration__

* some rows have no disease IDs
* a few NodeNorm mapping failures for OMIM IDs (several diff kinds): ~2.8%. 68 failures / (2401 unique values in column - 9 orphanet)
  * no NodeNorm mapping failures for MONDO IDs
* when rows have both OMIM and MONDO IDs, there are sometimes differences in NodeNorm mapping ("mismatches"). __In these cases, OMIM IDs were much more accurate__
* __MONDO IDs can be inaccurate__ - see the blue review boxes
  * VS it was much rarer to find an inaccurate OMIM ID mapping (found 1 case)


__Decision: Use OMIM ID column to generate NodeNorm values__

* less missing values
* seems to be more accurate (for successful NodeNorm mappings)

## Stats on rows removed during NodeNorming

This section prints the statistics on rows in the original data that were removed. 

(Uses variables generated during the previous section "Pre-NodeNorming")

<div class="alert alert-block alert-success">

**2025-03-28 data:** 

Genes: No rows removed due to lack of IDs for NodeNorming or NodeNorm mapping issues.

In [59]:
## partial put into parser (format): DONE

print("Gene Pre-NodeNorming\n")

## no gene IDs
print(f'{stats_no_gene_IDs["n_rows"]} row(s) with no gene IDs')

## no HGNC IDs: key column for NodeNorming
print(f'{n_rows_no_hgnc} row(s) with no HGNC IDs')

## HGNC NodeNorm issues: none, but showing anyways
print("\n")
print("HGNC NodeNorm mapping failures:")

print(f'IDs with no data in NodeNorm: {len(stats_hgnc_mapping_failures["nodenorm_returned_none"])}')
print(f'IDs with the wrong NodeNormed category: {len(stats_hgnc_mapping_failures["wrong_category"])}')
print(f'IDs with no label in NodeNorm: {len(stats_hgnc_mapping_failures["no_label"])}')

Gene Pre-NodeNorming

0 row(s) with no gene IDs
0 row(s) with no HGNC IDs


HGNC NodeNorm mapping failures:
IDs with no data in NodeNorm: 0
IDs with the wrong NodeNormed category: 0
IDs with no label in NodeNorm: 0


<div class="alert alert-block alert-success">

**2025-03-28 data:** 
    
__Diseases: many rows removed__ due to lack of IDs for NodeNorming or NodeNorm mapping issues.

In [60]:
## partial put into parser (format): DONE

print("Disease Pre-NodeNorming\n")

## no disease IDs
print(f'{stats_no_disease_IDs["n_rows"]} row(s) with no disease IDs '
      f'(= {stats_no_disease_IDs["n_names"]} unique diseases)')

## plus the rows that only lack OMIM IDs: key column for NodeNorming
print(f'+ {stats_mondo_only["n_rows"]} row(s) with no OMIM ID '
      f'(= {stats_mondo_only["n_names"]} unique diseases)')

## OMIM NodeNorm issues
print("\n")
print("OMIM NodeNorm mapping failures:")

print(f'{n_rows_no_data} row(s) for '
      f'{len(stats_OmOr_mapping_failures["nodenorm_returned_none"])} '
      f'IDs with no data in NodeNorm')

print(f'{n_rows_wrong_category} row(s) for '
      f'{len(stats_OmOr_mapping_failures["wrong_category"])} '
      f'IDs with the wrong NodeNormed category')

print(f'{n_rows_no_label} row(s) for '
      f'{len(stats_OmOr_mapping_failures["no_label"])} '
      f'IDs with no label in NodeNorm')

Disease Pre-NodeNorming

642 row(s) with no disease IDs (= 629 unique diseases)
+ 445 row(s) with no OMIM ID (= 278 unique diseases)


OMIM NodeNorm mapping failures:
41 row(s) for 39 IDs with no data in NodeNorm
27 row(s) for 26 IDs with the wrong NodeNormed category
3 row(s) for 3 IDs with no label in NodeNorm


<div class="alert alert-block alert-success">
    
__Totals__

In [61]:
## put into parser (format): DONE

n_rows_original = df.shape[0]
n_rows_removed = stats_no_disease_IDs["n_rows"] + stats_mondo_only["n_rows"] + \
                 n_rows_no_data + n_rows_wrong_category + n_rows_no_label
n_rows_after = n_rows_original - n_rows_removed

print(f"{n_rows_original} rows/records before Pre-NodeNorming\n")

print(f"{n_rows_removed} rows removed during Disease NodeNorming process\n")

print(f"{n_rows_after} rows/records left ({n_rows_after/n_rows_original:.1%})")

3659 rows/records before Pre-NodeNorming

1158 rows removed during Disease NodeNorming process

2501 rows/records left (68.4%)


## Adding NodeNorm data, removing rows

Using gene HGNC and disease OMIM/orphanet IDs for pre-NodeNorming

In [62]:
## put into parser (format): DONE

## Gene: assumes no missing values
df["gene_nodenorm_id"] = [hgnc_nodenorm_mapping[i]["primary_id"] for i in df["hgnc_id"]]
df["gene_nodenorm_label"] = [hgnc_nodenorm_mapping[i]["primary_label"] for i in df["hgnc_id"]]

df["disease_nodenorm_id"] = [OmOr_nodenorm_mapping[i]["primary_id"] 
                             if OmOr_nodenorm_mapping.get(i) 
                             else pd.NA
                             for i in df["disease_mim"]]

df["disease_nodenorm_label"] = [OmOr_nodenorm_mapping[i]["primary_label"] 
                                if OmOr_nodenorm_mapping.get(i) 
                                else pd.NA
                                for i in df["disease_mim"]]

In [63]:
df.head()

g2p_id gene_symbol     gene_mim     hgnc_id  \
0  G2P00001        HMX1  OMIM:142992   HGNC:5017   
1  G2P00002        SLX4  OMIM:613278  HGNC:23845   
2  G2P00003        ARG1  OMIM:608313    HGNC:663   
3  G2P00004         ATR  OMIM:601215    HGNC:882   
4  G2P00005       FANCB  OMIM:300515   HGNC:3583   

               previous_gene_symbols                          disease_name  \
0                         H6; NKX5-3  HMX1-related oculoauricular syndrome   
1  BTBD12; FANCP; KIAA1784; KIAA1987           SLX4-related Fanconi anemia   
2                                NaN              ARG1-related argininemia   
3            FRP1; MEC1; SCKL; SCKL1           ATR-related Seckel syndrome   
4              FAAP95; FAB; FLJ34064          FANCB-related Fanconi anemia   

   disease_mim disease_MONDO       allelic_requirement cross_cutting_modifier  \
0  OMIM:612109           NaN       biallelic_autosomal                    NaN   
1  OMIM:613951           NaN       biallelic_autosomal                    NaN   
2  OMIM:207800           NaN       biallelic_autosomal                    NaN   
3  OMIM:210600           NaN       biallelic_autosomal                    NaN   
4  OMIM:300514           NaN  monoallelic_X_hemizygous                    NaN   

   confidence  variant_consequence variant_types molecular_mechanism  \
0  definitive  absent gene product           NaN    loss of function   
1  definitive  absent gene product           NaN    loss of function   
2  definitive  absent gene product           NaN    loss of function   
3      strong  absent gene product           NaN    loss of function   
4  definitive  absent gene product           NaN    loss of function   

  molecular_mechanism_categorisation molecular_mechanism_evidence  \
0                           inferred                          NaN   
1                           inferred                          NaN   
2                           inferred                          NaN   
3                           inferred                          NaN   
4                           inferred                          NaN   

                                          phenotypes  \
0  HP:0000568; HP:0000589; HP:0000007; HP:0000482...   
1  HP:0000347; HP:0000007; HP:0001903; HP:0002984...   
2  HP:0000752; HP:0000737; HP:0000007; HP:0008339...   
3  HP:0000347; HP:0010230; HP:0001249; HP:0002750...   
4  HP:0000924; HP:0001871; HP:0001701; HP:0000083...   

                                        publications  \
0                       18423520; 25574057; 29140751   
1                                 21240275; 21240277   
2       10502833; 1598908; 7649538; 1463019; 2365823   
3                                                NaN   
4  16679491; 21910217; 36135330; 32106311; 307922...   

                                         panel comments  \
0       Developmental disorders; Eye disorders      NaN   
1                      Developmental disorders      NaN   
2                      Developmental disorders      NaN   
3  Developmental disorders; Skeletal disorders      NaN   
4      Developmental disorders; Skin disorders      NaN   

        date_of_last_review  \
0 2019-09-26 16:23:46+00:00   
1 2025-01-28 23:09:54+00:00   
2 2015-07-22 16:14:07+00:00   
3 2025-01-27 14:24:27+00:00   
4 2024-08-20 14:13:58+00:00   

                                      g2p_record_url gene_nodenorm_id  \
0  https://www.ebi.ac.uk/gene2phenotype/lgd/G2P00001    NCBIGene:3166   
1  https://www.ebi.ac.uk/gene2phenotype/lgd/G2P00002   NCBIGene:84464   
2  https://www.ebi.ac.uk/gene2phenotype/lgd/G2P00003     NCBIGene:383   
3  https://www.ebi.ac.uk/gene2phenotype/lgd/G2P00004     NCBIGene:545   
4  https://www.ebi.ac.uk/gene2phenotype/lgd/G2P00005    NCBIGene:2187   

  gene_nodenorm_label disease_nodenorm_id  \
0                HMX1       MONDO:0012802   
1                SLX4       MONDO:0013499   
2                ARG1       MONDO:0008814   
3                 ATR       MONDO:0008869   
4       

In [64]:
## put into parser (change in-place): DONE

df_only_nodenormed = df.dropna(subset=["gene_nodenorm_id", "gene_nodenorm_label", 
                                       "disease_nodenorm_id", "disease_nodenorm_label"],
                              ignore_index=True).copy()

In [65]:
## same! so it works as expected

df_only_nodenormed.shape

n_rows_after

(2501, 26)

2501

In [66]:
df_only_nodenormed.info(memory_usage="deep")

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2501 entries, 0 to 2500
Data columns (total 26 columns):
 #   Column                              Non-Null Count  Dtype              
---  ------                              --------------  -----              
 0   g2p_id                              2501 non-null   object             
 1   gene_symbol                         2501 non-null   object             
 2   gene_mim                            2501 non-null   object             
 3   hgnc_id                             2501 non-null   object             
 4   previous_gene_symbols               2248 non-null   object             
 5   disease_name                        2501 non-null   object             
 6   disease_mim                         2501 non-null   object             
 7   disease_MONDO                       117 non-null    object             
 8   allelic_requirement                 2501 non-null   object             
 9   cross_cutting_modifier              284 n

## Confidence values

**2024-04-15:**

Every row/record has 1 confidence value, representing how confident the curators are that "this gene has a causal role in this disease". The definitions of the possible values are provided [here (under G2P Confidence Category)](https://www.ebi.ac.uk/gene2phenotype/about/terminology). 


**CURRENT DEFINITIONS** (including in case they change later)

> **definitive**: The role of this gene in this particular disease has been repeatedly demonstrated in both the research and clinical diagnostic settings, and has been upheld over time (at least 2 independent publication over 3 years' time). No convincing evidence has emerged that contradicts the role of the gene in the specified disease. (previously labelled as confirmed) The strength of evidence within publications as well as their number and publication dates is taken into account. In practice, this usually means at least 4 publications over 5 years. Typically this will also include convincing bioinformatic or functional evidence of causation, making it very unlikely that this gene-disease association would ever be refuted.
>
>**strong**: The role of this gene as a monogenic cause of disease has been repeatedly and independently demonstrated providing very strong convincing evidence in humans and no conflicting evidence for this gene's role in this disease. (previously labelled as probable).
>
>**moderate**: There is moderate evidence in humans to support a casual role for this gene in this disease with no contradictory evidence. The body of evidence is not large (e.g possibly only one key paper) but appears convincing enough that the gene-disease pair is likely to be validated with additional evidence in the near future.
>
>**"limited"**: Little human evidence exists to support a casual role for this gene in this disease, but not all evidence has been refuted. For example, there may be a collection of rare missense variants in humans but without convincing functional impact, segregration data that could either arise by chance (e.g across one or two meioses) or does not implicate a single gene, or functional data without direct recapitulation of the phenotype. Overall, the body of evidence does not meet contemporary criteria for claiming a valid association with disease. The majority are probably false associations. (previously labelled as possible).
>
>**"disputed"**: "Although evidence has been reported, other evidence of equal weight disputes the claim."
>
>**"refuted"**: "There has been an assertion of a gene-disease association in the literature, but new valid evidence has arisen that refutes the entire original body of evidence."

<div class="alert alert-block alert-success">

**2024-04-15:**

My thinking is...
1. rows with **"refuted"** and **"disputed"** values **should not be used to create edges for Translator**, because there's strong evidence that there ISN'T an association (negation) based on the definitions. 
2. rows with **"limited"** confidence can be kept because I interpret the definition as saying there is AN association - it's just not causal (as far as we know) and it's unclear how "real"/important it is. So these rows should have a predicate weaker than "causes"/"contributes to" -> **using "related to" for now**. 
3. keep rows with **"moderate", "strong", "definitive"** values, because there's moderate-definitive evidence that a gene DOES HAVE a causal role in this disease -> **"using "causes" for now**

Other possible data modeling: use subject_form_or_variant_qualifier *genetic_variant_form*. Okay because every row has an allelic_requirement, and those [terms](https://www.ebi.ac.uk/gene2phenotype/about/terminology) are for mutations.

<div class="alert alert-block alert-danger">

Data-modeling notes: options for gene-disease associations are confusing 
* can "causes / contributes to" be used? Maybe it makes more sense to use them with qualifiers on gene/protein (form or variant, aspect)...but are we allowed to use qualifiers here?
* what's the diff between "associated with" and "genetically associated with"? 
* "gene associated with condition" is child of "genetically associated with", but seems to be more general - basically a "related to". Also would look weird in UI, right? 

In [69]:
## NodeNormed data only
df_only_nodenormed["confidence"].value_counts()

confidence
definitive    1614
strong         542
limited        255
moderate        89
refuted          1
Name: count, dtype: int64

In [68]:
## all data (before NodeNorming)
df["confidence"].value_counts()

confidence
definitive    2047
strong         853
limited        518
moderate       240
refuted          1
Name: count, dtype: int64

**2025-03-28 data:** 
No "disputed" values, only 1 "refuted" row to remove

### Removing rows + stats

In [72]:
## calculate stats before removing

n_rows_before = df_only_nodenormed.shape[0]
n_rows_refuted = df_only_nodenormed[df_only_nodenormed["confidence"] == "refuted"].shape[0]
n_rows_disputed = df_only_nodenormed[df_only_nodenormed["confidence"] == "disputed"].shape[0]

In [73]:
## remove rows, calculate stats after

df_only_nodenormed = df_only_nodenormed[~ df_only_nodenormed["confidence"].isin(["refuted", "disputed"])]
n_rows_after = df_only_nodenormed.shape[0]

In [75]:
## Print stats

print(f"Number of rows before: {n_rows_before}\n")

print(f"Removing rows based on confidence:")
print(f"{n_rows_refuted}: 'refuted'")
print(f"{n_rows_disputed}: 'disputed'")
print("\n")

print(f"Number of rows after: {n_rows_after}")

Number of rows before: 2501

Removing rows based on confidence:
1: 'refuted'
0: 'disputed'


Number of rows after: 2500


## Generating documents

In [67]:
## code chunk for testing parts of inner code

for row in df.itertuples(index=False):
    if pd.notna(row.previous_gene_symbols):
        [i.strip() for i in row.previous_gene_symbols.split(";")]
        break

['ISK', 'JLNS2', 'LQT5', 'MINK']

In [68]:
## put into parser (format): DONE
##   don't save in array, yield each document instead

## GENERATING DOCS, saving in array
documents = []

## using itertuples because it's faster, preserves datatypes
for row in df_only_nodenormed.itertuples(index=False):
    ## simple assignments
    document = {
        "_id": row.g2p_id,
        "subject": {
            "id": row.gene_nodenorm_id,
            "label": row.gene_nodenorm_label,
            "original_subject": row.hgnc_id,
            "original_data": {
                "hgnc_symbol": row.gene_symbol,
                "hgnc": row.hgnc_id,
            },
            "type": "Gene"
        },
        "association": {
            "g2p_record_id": row.g2p_id,
            "g2p_record_url": row.g2p_record_url,
            "allelic_requirement": row.allelic_requirement,
            "confidence": row.confidence,
            "molecular_mechanism": row.molecular_mechanism,
            "molecular_mechanism_categorisation": row.molecular_mechanism_categorisation,
            "g2p_panels": [i.strip() for i in row.panel.split(";")],
            "date_of_last_review": str(row.date_of_last_review)
        },
        "object": {
            "id": row.disease_nodenorm_id,
            "label": row.disease_nodenorm_label,
            ## currently, after NodeNorming, no NAs in OMIM/orphanet column
            "original_object": row.disease_mim,
            "original_data": {
                ## not putting disease_mim here: parsing into separate OMIM and orphanet fields 
                "name": row.disease_name,
            },
            "type": "Disease"
        }
    }    
    ## only create field if value is not NA
    ##   if value is NA, list comprehension with split won't work
    ## Gene
    if pd.notna(row.gene_mim):
        document["subject"]["original_data"]["omim"] = row.gene_mim
    if pd.notna(row.previous_gene_symbols):
        document["subject"]["original_data"]["previous_gene_symbols"] = \
        [i.strip() for i in row.previous_gene_symbols.split(";")]
        
    ## Association
    if pd.notna(row.cross_cutting_modifier):
        document["association"]["cross_cutting_modifiers"] = [i.strip() for i in row.cross_cutting_modifier.split(";")]
    if pd.notna(row.variant_consequence):
        document["association"]["variant_consequences"] = [i.strip() for i in row.variant_consequence.split(";")]
    if pd.notna(row.variant_types):
        document["association"]["variant_types"] = [i.strip() for i in row.variant_types.split(";")]
    ## uses diff delimiters, could do more parsing
    if pd.notna(row.molecular_mechanism_evidence):
        document["association"]["molecular_mechanism_evidence"] = [i.strip() for i in row.molecular_mechanism_evidence.split("&")]
    if pd.notna(row.phenotypes):
        document["association"]["phenotypes"] = [i.strip() for i in row.phenotypes.split(";")]
    if pd.notna(row.publications):
        document["association"]["pmids"] = [i.strip() for i in row.publications.split(";")]
    if pd.notna(row.comments):
        document["association"]["curator_comments"] = row.comments
     
    ## Disease
    ## disease_mim: create field depending on whether OMIM or orphanet    
    if pd.notna(row.disease_mim):
        if row.disease_mim.startswith("orphanet"):
            document["object"]["original_data"]["orphanet"] = row.disease_mim
        elif row.disease_mim.startswith("OMIM"):
            document["object"]["original_data"]["omim"] = row.disease_mim
    if pd.notna(row.disease_MONDO):
        document["object"]["original_data"]["mondo"] = row.disease_MONDO
    
    documents.append(document)

## Checking documents

In [69]:
len(documents)

2499

In [70]:
## code chunk for finding rows -> then look up corresponding doc by idx
# df_only_nodenormed[df_only_nodenormed["disease_mim"].str.contains("orphanet", na=False)]


# df_only_nodenormed[df_only_nodenormed["previous_gene_symbols"].isna()]
# df_only_nodenormed[df_only_nodenormed["disease_MONDO"].notna()]

In [71]:
pprint(documents[416])

# documents[416]

{'_id': 'G2P00116',
 'association': {'allelic_requirement': 'biallelic_autosomal',
                 'confidence': 'definitive',
                 'date_of_last_review': '2015-07-22 16:14:16+00:00',
                 'g2p_panels': ['Developmental disorders'],
                 'g2p_record_id': 'G2P00116',
                 'g2p_record_url': 'https://www.ebi.ac.uk/gene2phenotype/lgd/G2P00116',
                 'molecular_mechanism': 'loss of function',
                 'molecular_mechanism_categorisation': 'inferred',
                 'phenotypes': ['HP:0003281',
                                'HP:0001903',
                                'HP:0003593',
                                'HP:0001425',
                                'HP:0000007'],
                 'pmids': ['19412178'],
                 'variant_consequences': ['absent gene product']},
 'object': {'id': 'MONDO:0008785',
            'label': 'sideroblastic anemia 2',
            'original_data': {'name': 'SLC25A38-related anemia

## Parser notes

Fine to use raise/assert in parser (raise is technically better programming behavior: https://realpython.com/python-assert-statement/#understanding-common-pitfalls-of-assert)


My notes on parser:
* adding prefixes to gene/disease IDs is good for pre-NodeNorming steps
* keeping diff gene/disease ID namespaces as separate fields right now is good for current BTE/x-bte-annotation system
  * Also, original subject will always be HGNC, original object will always be disease OMIM with current code


My notes on syntax:
* use `yield` when you want to "return" within a "for loop" (return only happen once, then exit for-loop/function execution)
  * that's what it's used in main execution, when you're iterating over csv rows to generate documents
* use `yield from {function}` to get the data from a generator (created by `yield` being used the function)